### ISSUE TITLE
Graph Autoencoder Feature Selection with hnswlib

### ISSUE THREAD

**Issue by jlevy44** [pyg-team/pytorch_geometric#470]
Hey there. I've been interested in implementing: https://arxiv.org/pdf/1801.02251.pdf

Do you think this is possible with random minibatches?

I know it doesn't necessarily match the formalism introduced by pytorch geometric, but I wanted to use hnswlib to construct the weighted adjacency matrix (using cosine similarity), then calculate the laplacian, then embed the data X without graph structure to form Z, and then calculate tr(ZLZ^T) and optimize that, but for each batch.

Is this possible? What do you think is the easiest way to get there? I feel that the main issue would be making the calculation of the Laplacian work with SBGD, which I feel could be an impossibility.

**Comment by rusty1s** [comment]
Why do you think the Laplacian calculation is impossible? Given that you have a sparse diagonally stacked graph obtained from modified knn_graph, you can easily compute the Laplacian. You need to convert your data into a dense setting using `to_dense_adj` before calculate Z L Z^T, and I am not sure if PyTorch provides a dense trace functionality. Otherwise, you need to implement ihr by yourself.

**Comment by jlevy44** [comment]
I suppose not impossible. The original implementation does not optimize over randomized minibatches. I think calculating the laplacian over each minibatch would be different than L over the entire dataset. This seems like you’d throw out important adjacency information for every minibatch, unless I’m mistaken or reading it the wrong way. To truly preserve the geometric information, wouldn’t one want to use the laplacian from all of the datapoints rather than each batch if using knn?

**Comment by rusty1s** [comment]
I think I do not fully understand. You have an adjacency matrix A (diagonally stacked using `knn_graph`, so no connections between data points in different examples), and compute `L = D - A`. What do I miss?

**Comment by jlevy44** [comment]
I think the goal is to not diagonally stack A. I’d want to calculate A outside of any batch. If memory were not a problem, I’d want to do this with one batch over the entire dataset. In the end, My goal is to use one graph, and it’s associated laplacian, as a regularizer, not by batch. Similar to UMAP, the goal is global geometric preservation, so if examples were batches, I would somehow want a batch level computation of ZLZ^T over the entire dataset. But by splitting up into batches, while you get Z, calculating L in each batch throws out a more global context.

**Comment by rusty1s** [comment]
I must admit that I haven't read the paper, so I might not be of any help here.

So you want to produce "one giant graph", but process it in a mini-batch scenario, *e.g.* by filtering `L` and `Z`? I guess the problem is then to compute a mini-batch version of `Z`, and use it to compute `ZLZ^T`? I think you are right when saying that you lose out on global context that way, but you could argue that this is normal with SGD (and may be even beneficial).

**Comment by jlevy44** [comment]
Mini-batch z is definitely doable. So one need not compute Z while computing L.

Would you say that it is reasonable to filter L for the included indices for a given minibatch?

The problem with creating L in a mini-batch situation is that how could the produced global affinity graph even be remotely comparable to the mini-batch one? 

I'm just considering the pros and cons of each approach. The ideal case might be running one batch per epoch on cpu, which would be much slower. I think the creators did not have in mind how many dimensions the data might scale to.

**Comment by rusty1s** [comment]
Yes, this sounds indeed really strange to just use random mini-batches. I think it may be reasonable to create mini-batches based on cosine similarity? This way you at least preserve locality.

**Comment by jlevy44** [comment]
Sure, that could be an approach. The goal of this algorithm though is to find features that preserve globally.
